### ISSUE TITLE
Possible code improvements

### ISSUE THREAD

**Issue by hackerbirds** [QubesOS/qubes-notification-proxy#3]
Hey, I got curious and decided to read the code of this repository, and I noticed a few things that might be improved in the Rust code. It's completely understandable if "it's good enough" as it is though, so feel free to close this issue if it's not a priority.

Full disclaimer, 
1) I can't pretend I understand how or where the code works exactly, so I won't comment on any of the logic behind what's being done. 
2) I can't build/test this crate because I'm missing the `qubes-pure` library, so I couldn't test any of the suggestions I'm going to make
3) There were probably some design decisions being made in the code that I'm unaware of, so please let me know if something I pointed out is intentional

### Tokio *might* be overkill?

I noticed that `tokio` is used in a local/single-threaded way to use `!Send` tasks on the local thread. While this is completely fine, `tokio` is mostly designed around multi-threaded environments and might be "overkill" for this use.

`smol` instead of `tokio` could be a sound choice if you want async in a single-threaded environment, and would most likely reduce the total number of dependencies. However I must point out that I have never used `smol` myself so I can't attest to its quality or ease-of-use. I do know that it is a single-threaded async executor, so you can pass references/`!Send` things into tasks, and it is more "modular" than `tokio` in that you can import only the async primitives you need and nothing else.

### `String`/`&str`
I see some `&str` being used in function arguments, but they seem to only be used in places where you want to pass a `String` in all the real world cases (i.e. outside of tests). 

For instance, `NotificationEmitter::send_notification(&self, Notification)` receives a Notification which contains `String`s inside, but the sanitizing function reads a `&str` in the input and returns `String` in the output! And then you dereference this sanitized `String` back into `str`, then reference it into `&str`, without using the ownership of `str` in the rest of the function's scope. This results in a lot of the strings looking like `&*some_string`. 

https://github.com/QubesOS/qubes-notification-proxy/blob/main/src/lib.rs#L583
https://github.com/QubesOS/qubes-notification-proxy/blob/main/src/lib.rs#L613

You are unnecessarily dereferencing things here. What you're doing doesn't have any real cost, but it doesn't play well with Rust's ownership principles. I suggest you pass `String`, an owned type, into the function arguments that will mutate/play around with it, and return another `String`. More specifically in what I saw, changing `sanitize_str` and `Notifications::notify` to take `String` instead of `&str`, because it is going to use the `String` and therefore should also consume its ownership in this case.

In the rare events where you might somehow need to use the same `String` twice, and you can't pass a reference of it, then `.clone()` is much simpler (that doesn't appear to be needed at all in what I've seen though).

If you must convert a `String` into an `&str`, like when you want to `match` string slices from a `String` like you're doing for instance in `lib.rs:33``, instead of passing `&*a_string` in the match I suggest using `a_string.as_str()`.

https://github.com/QubesOS/qubes-notification-proxy/blob/main/src/lib.rs#L331 

Lastly, `Server::get_capabilities` returns a `Vec<String>`, but it contains string slices that are already known at compile time. Every time you're calling `get_capabilities`, you're turning those `&'static str` into `String` at run time. In this case it's fine to return a `Vec<&'static str>`. The reference has a static lifetime here and won't drop or anything like that. That said I don't think this specific function is used in the code.

### In the near future you could remove `serde` as a dependency

`bincode` version 1 used `serde` as a depedency to serialize/deserialize into bytes, but that will become optional in `bincode` 2.0.0. Instead in `bincode` 2, you can derive the `Encode`/`Decode` traits instead of having to use `serde`'s `Serialize`/`Deserialize`. Since you use `serde` exclusively for this basic case, then you might want to switch to `Encode`/`Decode` in the future.

Bincode 2.0.0 isn't stabilised *yet*, so I wouldn't make that change now. But it's something that should be considered in the future. Less dependencies = less supply chain issues etc.

### Maybe?: Reduce the use of `tokio::task::spawn_local`

`notification-proxy-server::client_server` spawns 3 local tasks that wait for `Future`s from different streams.

It might be possible to simplify the code here using `select!` https://tokio.rs/tokio/tutorial/select. It allows you to wait on all those different `while ... .await {}` receivers in one.

It could be used something like this (I didn't test it):

```rust
loop {
    tokio::select! {
        item = server_name_owner_changed.next() => { ... },
        item = closed_stream.next() => { ... },
        item = invoked_stream.next() => { ... },
        else => {
            // All the three channels were closed/returned `None`
        }
    }
}
```

The `loop` at `notification-proxy-server::client_server::115` might also benefit from this by using `msg = stdin.read_u32_le() => { ... }`, though not without some significant structural changes.

Note the `loop` here--`select!` is built to return as soon as one of the futures return. If you have one of the streams return something, `select!` will return regardless of the state of the other streams. It is common to use `loop` on top of `select!`, and `tokio` guarantees that if `select!` returns, the other streams will be fine/unchanged. I think doing this also has the benefit of not relying on `move { }` closures, so you could be using a single `stdout` and `emitter` directly for all of the branches in the select!, without needing to `Rc`/`.clone()` it every time for each branch like it's currently being done.

---

There are probably other tiny improvements you could make, but I can't play around the code too much because I don't know how to build/test it properly.

**Comment by hackerbirds** [comment]
> The loop at notification-proxy-server::client_server::115 might also benefit from this by using msg = stdin.read_u32_le() => { ... }, though not without some significant structural changes.

Actually, it is *necessary* even to do it, since you're most likely going to want to `loop` the `select!`. If you do that, then you have to make all the code after it reachable by also including it in the `select!`.

**Comment by DemiMarie** [comment]
#4 makes it possible to build this outside of Qubes OS, by disabling the `unicode` feature.  It also allows building with the runtime versions of the libraries, rather than the development versions.  With this change, the code should build out of the box on any Qubes system, and it should build anywhere if `--no-default-features` is passed.

**Comment by DemiMarie** [comment]
Is all of this code cancellation-safe?  I’m not sure if `select!` is safe to use here.

Tokio is _massive_ overkill for this use-case.  I would much rather use something like `smol` or even `calloop`.  There are two other questions:

1. Does Fedora ship this as a package?  Using packages from `crates.io` directly is problematic for Qubes OS.  It looks like `calloop` is shipped but `smol` is not.
2. How do I integrate a chosen D-Bus library with these crates?
3. Should I use zbus, or should I pick a different library?  All else equal, I would much rather use a libdbus binding than a pure-Rust implementation, as it would be simpler and easier to audit.  (The system libdbus is used by too much other code and so must be trusted anyway.)

**Comment by hackerbirds** [comment]
> #4 makes it possible to build this outside of Qubes OS, by disabling the `unicode` feature. It also allows building with the runtime versions of the libraries, rather than the development versions. With this change, the code should build out of the box on any Qubes system, and it should build anywhere if `--no-default-features` is passed.

Thanks!

> Is all of this code cancellation-safe? I’m not sure if `select!` is safe to use here.

It appears that `zbus` uses `futures::stream::Stream`, [which would be cancellation safe](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety), so using theNotificationEmitter streams in `select!` seem fine. 

However, the `read_u32_le()` loop on `stdin` isn't cancellation safe, so that can't be used in a `select!` as it is. The link I put above suggests that `read()` and `read_buf()` are though, so if that can be used instead of `read_u32_le()`+`read_exact()`, that could work.

> Does Fedora ship this as a package? 

Hah, didn't think of that. No idea. Maybe `smol` won't work then for now.

> How do I integrate a chosen D-Bus library with these crates?
> Should I use zbus, or should I pick a different library? All else equal, I would much rather use a libdbus binding than a pure-Rust implementation, as it would be simpler and easier to audit. (The system libdbus is used by too much other code and so must be trusted anyway.)

Honestly I cannot help with that sorry. I've never worked with D-Bus before. I have no knowledge on what would have to be done. I found this `libdbus` binding library though at the bottom `zbus`'s README, which also happens to have `tokio` support. https://github.com/diwic/dbus-rs/. But I can't stress enough that I have tested neither of these crates.

**Comment by hackerbirds** [comment]
I tested #4, it compiles just fine for me now. If that's ok, I'll make a PR soon where I make tiny (mostly readability) improvements without affecting the code logic or anything else as a start

**Comment by hackerbirds** [comment]
> However, the `read_u32_le()` loop on `stdin` isn't cancellation safe, so that can't be used in a `select!` as it is. The link I put above suggests that `read()` and `read_buf()` are though, so if that can be used instead of `read_u32_le()`+`read_exact()`, that could work.

~~Do you think it could be possible to use `take(MAX_MESSAGE_SIZE as u64)` (https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#method.take) on `stdin` in combination with `read(&mut buf)` to make this portion of the code cancellation safe? This would limit the size of messages received from `stdin` without having to use `read_u32_le()`. If there's a message larger than that being passed, then from my understanding `read` will return `Ok(0)`, which you could use to `panic!` because the message is too large. Deserialisation would probably fail too in that scenario.~~

~~If that works, I think everything else is cancellation safe, and `select!` could be used.~~

--- 
Alright, I was wrong. `take` limits the entire stdin, not just the read portion. 

Looking deeper and experimenting around though I found that you could use framing. There is a length prefix codec in `tokio-util` (you need the `codec` feature) that seems to do what is desired, and returns a FramedRead which implements Stream, so using .next() on it is cancellation safe. https://docs.rs/tokio-util/latest/tokio_util/codec/length_delimited/index.html

Here's how it could work. As always, further tests needed:

```rust
use futures_util::StreamExt;
use tokio::select;
use tokio_util::codec::LengthDelimitedCodec;

#[tokio::main]
async fn main() {
    let max_msg_size: usize = 10;
    let mut stdin = LengthDelimitedCodec::builder()
        .length_field_type::<u32>()
        .little_endian()
        .max_frame_length(max_msg_size)
        .new_read(tokio::io::stdin());

    loop {
        select! {
            Some(Ok(line)) = stdin.next() => {
                // ...
            },
            else => {
                panic!("Error happened");
            }
        }
    }
}
```

The main downside here being the extra dependency needed for a `tokio` crate. Though implementing the codec itself doesn't seem too hard
